src/hooks/useQueryState.js

```diff
@@ -0,0 +1,163 @@
+   * This function pulls the value of the key from the url
+   *   and updates the state with it (if changed)
+   *
+   * This is a useLayoutEffect rather than useEffect in order to prioritize loadURLParams;
+   *   this is necessary for preventing the race condition where (for ex.) the user clicks 'back'
+   *   twice quickly and the first 'back' sets the URL params before the second 'back' can
+   *   load the accurate new URL param. This is still not totally perfect but only breaks when
+   *   a user really messes with the navigation significantly and quickly.
+   * Comparison of the two effects: https://kentcdodds.com/blog/useeffect-vs-uselayouteffect
+   */
+  useLayoutEffect(() => {
+    const loadURLParams = async () => {
+      const urlParams = new URLSearchParams(history.location.search);
+      let urlValue = urlParams.get(key);
+
+      switch (type) {
+        // using boolean key -> load true if present, false if absent from url
+        case types.Boolean:
+          setState(urlParams.has(key));
+          return;
+
+        // different logic if we are working with an array than with single value
+        // because setState will not do deep array comparison
+        case types.Array:
+          // if the same value (or unset and state array is empty) -> do nothing
+          // this check is necessary because [] -> [] are different memory objects
+          if (urlValue === state.toString() || (!urlValue && state.length === 0)) return;
+          else if (!urlValue) setState(defaultVal);
+          else setState(urlValue.split(','));
+          break;
+
+        case types.SingleValue:
+          setState(urlValue ?? defaultVal);
+          break;
+
+        default:
+          // should never be reached, type must always be one of the preceeding
+          throw new Error(`missing case for type ${type}`);
+      }
+    };
+    loadURLParams();
+    /*********************** ESLint Disable Justification: *******************************/
+    // ESLint wants to include `defaultVal, key, state` as dependencies but to prevent
+    // extra loops of ~ url setting state setting the url ~ we do not include.
+    // It makes good sense that the only time we should pull the URL is the URL changes
+    /*************************************************************************************/
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [history.location.search]);
+
+  /**
+   * Callback function to allow user component to setQueryState just like one would use setState
+   */
+  const setQueryState = useCallback((val) => {
+    // this check prevents unnecessary array setting
+    // ([] -> [] are technically different memory objects)
+    // this is safe to deep compare because the array value should not have very many values
+    // (ex. 'filters' array has a few different possible filter values)
+    if (Array.isArray(defaultVal) && !JSON.stringify(state) === JSON.stringify(val)) return;
```

Comment:
Is there a reason you're checking the type of `defaultVal` instead of referencing `type` like you do elsewhere in this hook?

Also, `JSON.stringify()` will work but it's one of the slowest options. Perhaps `Array.every` is better here?
```suggestion
    if (type === types.Array && !state.every((elem, index) => elem === val[index])) return;
```